crates/types/src/lib.rs

```diff
@@ -69,8 +69,117 @@ impl fmt::Display for WasmType {
             WasmType::F32 => write!(f, "f32"),
             WasmType::F64 => write!(f, "f64"),
             WasmType::V128 => write!(f, "v128"),
-            WasmType::ExternRef => write!(f, "externref"),
-            WasmType::FuncRef => write!(f, "funcref"),
+            WasmType::Ref(rt) => write!(f, "{}", rt),
+            WasmType::Bot => write!(f, "bot"),
+        }
+    }
+}
+
+/// WebAssembly reference type -- equivalent of `wasmparser`'s RefType
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct WasmRefType {
+    pub nullable: bool,
+    pub heap_type: WasmHeapType,
+}
+
+pub const WASM_EXTERN_REF: WasmRefType = WasmRefType {
+    nullable: true,
+    heap_type: WasmHeapType::Extern,
+};
+
+pub const WASM_FUNC_REF: WasmRefType = WasmRefType {
+    nullable: true,
+    heap_type: WasmHeapType::Func,
+};
+
+impl TryFrom<wasmparser::RefType> for WasmRefType {
+    type Error = WasmError;
+    fn try_from(
+        wasmparser::RefType {
+            nullable,
+            heap_type,
+        }: wasmparser::RefType,
+    ) -> Result<Self, Self::Error> {
+        Ok(WasmRefType {
+            nullable,
+            heap_type: WasmHeapType::try_from(heap_type)?,
+        })
+    }
+}
+
+impl From<WasmRefType> for wasmparser::RefType {
+    fn from(
+        WasmRefType {
+            nullable,
+            heap_type,
+        }: WasmRefType,
+    ) -> wasmparser::RefType {
+        wasmparser::RefType {
+            nullable,
+            heap_type: wasmparser::HeapType::from(heap_type),
+        }
+    }
+}
+
+impl fmt::Display for WasmRefType {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &WASM_EXTERN_REF => write!(f, "externref"),
+            &WASM_FUNC_REF => write!(f, "funcref"),
+            WasmRefType {
+                heap_type,
+                nullable,
+            } => {
+                if *nullable {
+                    write!(f, "(ref null {})", heap_type)
+                } else {
+                    write!(f, "(ref {})", heap_type)
+                }
+            }
+        }
+    }
+}
+
+/// WebAssembly heap type -- equivalent of `wasmparser`'s HeapType
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub enum WasmHeapType {
+    Bot,
+    Func,
+    Extern,
+    Index(u32),
+}
+
+impl TryFrom<wasmparser::HeapType> for WasmHeapType {
+    type Error = WasmError;
+    fn try_from(ht: wasmparser::HeapType) -> Result<Self, Self::Error> {
+        use wasmparser::HeapType::*;
+        match ht {
+            Bot => Ok(WasmHeapType::Bot),
+            Func => Ok(WasmHeapType::Func),
+            Extern => Ok(WasmHeapType::Extern),
+            Index(i) => Ok(WasmHeapType::Index(i)),
+        }
+    }
+}
```

Comment:
Why is this a `TryFrom` impl instead of a `From` impl? It appears to be infallible.